### ISSUE TITLE
Guildline for Architecture which can gain from FP16

### ISSUE THREAD

**Issue by melgor** [NVIDIA/apex#196]
First of all, thanks for Apex 1.0. It is super easy now to check if FP16 works well in some architecture.
Then I decided to test some architecture on 2080Ti:

[Se101-resnext](https://github.com/Cadene/pretrained-models.pytorch/blob/master/pretrainedmodels/models/senet.py#L434):
FP32: 18 images/s
FP16: 21 images/s

[Se-Res101](https://github.com/Cadene/pretrained-models.pytorch/blob/master/pretrainedmodels/models/senet.py#L401):
FP32: 23 mages/s
FP16: 29 images/s

[Res101](https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py#L205):
FP32: 33 images/s
FP16: 66 images/s (exactly 2x more)

[Res152](resnet152):
FP32: 20.5 images/s
FP16: 40 images/s

[ResNext101-32x4d](https://github.com/Cadene/pretrained-models.pytorch/blob/master/pretrainedmodels/models/resnext_features/resnext101_32x4d_features.py)
FP32: 23 images/s
FP16: 33 images/s 

So in general ResNet 2x faster, ResNext 40% faster, Se-Res101 25%. Se-ResNext 16%.
Adding both ResNext or Se or both make speedup much smaller. 

I read that Upsumpling make FP16 slow, strided/dilated convolution. 
In ResNext maybe 'Groups-Conv' make it slower?
In SE maybe SEModule (AdaptivePool-maybe changing to AvgPool would be fine?, Sigmoid)?

Then the Guildline would be usefull which would point which type of connection are supported by FP16 == the gain is> 50%.

What do you think about that?
@ngimel look like you have a lot of knowlednge about CuDnn and TesnorCore so you maybe helpfull there.

**Comment by ngimel** [comment]
For Se101-resnext and se-Res101, depending on which version of pytorch you are using, you might have to replace adaptiveAveragePooling with `mean`. Latest pytorch nightlies have this fix, but 1.0.1 does not. Diff below (for Remi Cadene's pretrained-models.pytorch repo, I don't know which version you are using but it should be similar)
```
diff --git a/pretrainedmodels/models/senet.py b/pretrainedmodels/models/senet.py
index 20bd122..e6ec14d 100644
--- a/pretrainedmodels/models/senet.py
+++ b/pretrainedmodels/models/senet.py
@@ -86,7 +86,7 @@ class SEModule(nn.Module):
 
     def __init__(self, channels, reduction):
         super(SEModule, self).__init__()
-        self.avg_pool = nn.AdaptiveAvgPool2d(1)
+#        self.avg_pool = nn.AdaptiveAvgPool2d(1)
         self.fc1 = nn.Conv2d(channels, channels // reduction, kernel_size=1,
                              padding=0)
         self.relu = nn.ReLU(inplace=True)
@@ -96,7 +96,7 @@ class SEModule(nn.Module):
 
     def forward(self, x):
         module_input = x
-        x = self.avg_pool(x)
+        x = x.view(x.size(0), x.size(1), -1).mean(-1).view(x.size(0), x.size(1), 1,1)#self.avg_pool(x)
         x = self.fc1(x)
         x = self.relu(x)
         x = self.fc2(x)
```
Also, there was a bug in pytorch in algo selection for grouped convolution which was fixed may be a month back, so latest nightlies would be ok, but 1.0.1 probably has it. 
You are right that Group convolutions usually provide limited speed-up, that's why you are seeing smaller speed-up on resnext derived models.

**Comment by rwightman** [comment]
@ngimel To confirm, if I'm seeing a performance decrease for a resnext50 network with amp enabled, is it likely due to the mentioned group conv algo selection bug? Dual Turing GPU, Pytorch 1.0.1 Conda install.

My Resnet50 baseline for the same training code is approx 500 ips w/o AMP and 950 ips w/ AMP. ResNext50 is 400 ips w/o AMP and decreases to 260 ips w AMP.
